doc/man/lttng.1
.TH "LTTNG" "1" "May 13th, 2014" "" ""

.SH "NAME"
lttng \- LTTng 2.x tracer control command line tool

.SH "SYNOPSIS"

.PP
lttng [OPTIONS] <COMMAND>
.SH "DESCRIPTION"

.PP
The LTTng project aims at providing highly efficient tracing tools for Linux.
Its tracers help track down performance issues and debug problems
involving multiple concurrent processes and threads. Tracing across multiple
systems is also possible.

The \fBlttng\fP command line tool from the lttng-tools package is used to control
both kernel and user-space tracing. Every interaction with the tracer should
be done by this tool or by the liblttng-ctl library provided by the lttng-tools
package.

LTTng uses a session daemon (lttng-sessiond(8)), acting as a tracing registry,
which allows you to interact with multiple tracers (kernel and user-space)
inside the same container, a tracing session. Traces can be gathered from the
kernel and/or instrumented applications (lttng-ust(3)). Aggregating and reading
those traces is done using the babeltrace(1) text viewer.

We introduce the notion of \fBtracing domains\fP which is essentially a type of
tracer (kernel, user space, JUL, LOG4J or Python for now). In the future, we
could see more tracer like for instance an hypervisor. For some commands,
you'll need to specify on which domain the command operates (\-u, \-k, \-l, \-j
or \-p). For instance, the kernel domain must be specified when enabling a
kernel event.

In order to trace the kernel, the session daemon needs to be running as root.
LTTng provides the use of a \fBtracing group\fP (default: tracing). Whomever is
in that group can interact with the root session daemon and thus trace the
kernel. Session daemons can co-exist, meaning that you can have a session daemon
running as Alice that can be used to trace her applications along side with a
root daemon or even a Bob daemon. We highly recommend starting the session
daemon at boot time for stable and long term tracing.

Each user-space application instrumented with lttng-ust(3) will automatically
register with the root session daemon and its user session daemon. This allows
each daemon to list the available traceable applications and tracepoints at any
given moment (See the \fBlist\fP command).
.SH "OPTIONS"

.PP
This program follows the usual GNU command line syntax with long options starting with
two dashes. Below is a summary of the available options.
.PP

.TP
.BR "\-h, \-\-help"
Show summary of possible options and commands.
.TP
.BR "\-V, \-\-version"
Show version.
.TP
.BR "\-v, \-\-verbose"
Increase verbosity.
Three levels of verbosity are available which are triggered by putting additional v to
the option (\-vv or \-vvv)
.TP
.BR "\-q, \-\-quiet"
Suppress all messages (even errors).
.TP
.BR "\-g, \-\-group NAME"
Set unix tracing group name. (default: tracing)
.TP
.BR "\-n, \-\-no-sessiond"
Don't automatically spawn a session daemon.
.TP
.BR "\-\-sessiond\-path PATH"
Set session daemon full binary path.
.TP
.BR "\-\-list\-options"
Simple listing of lttng options.
.TP
.BR "\-\-list\-commands"
Simple listing of lttng commands.
.TP
.BR "\-m, \-\-mi TYPE
Machine interface

TYPE supported: XML

Machine interface (MI) mode converts the traditional pretty printing to a
machine output syntax. MI mode provides a format change-resistant way to access
information generated via the lttng command line.

When using MI mode, the data is printed on \fBstdout\fP. Error and warning are
printed on \fBstderr\fP with the pretty print default format.

If any errors occur during the execution of a command, the return value of the
command will be different than zero. In this case, lttng does NOT guarantee the
syntax and data validity of the generated MI output.

For XML output type, a schema definition (XSD) file used for validation can be
found under src/common/mi_lttng.xsd

.SH "COMMANDS"

.PP
\fBadd-context\fP [OPTIONS]
.RS
Add context to event(s) and/or channel(s).

A context is basically extra information appended to a channel. For instance,
you could ask the tracer to add the PID information for all events in a
channel. You can also add performance monitoring unit counters (perf PMU) using
the perf kernel API.

For example, this command will add the context information 'prio' and two per-CPU
perf counters (hardware branch misses and cache misses), to all events in the trace
data output:

.nf
# lttng add-context \-k \-t prio \-t perf:cpu:branch-misses \\
		\-t perf:cpu:cache-misses
.fi

Please take a look at the help (\-h/\-\-help) for a detailed list of available
contexts.

Perf counters are available as per-CPU ("perf:cpu:...") and per-thread
("perf:thread:...") counters. Currently, per-CPU counters can only be
used with the kernel tracing domain, and per-thread counters can only be
used with the UST tracing domain.

If no channel is given (\-c), the context is added to all channels that were
already enabled. If the session has no channel, a default channel is created.
Otherwise the context will be added only to the given channel (\-c).

If \fB\-s, \-\-session\fP is omitted, the session name is taken from the .lttngrc
file.

.B OPTIONS:

.TP
.BR "\-h, \-\-help"
Show summary of possible options and commands.
.TP
.BR "\-s, \-\-session NAME"
Apply on session name.
.TP
.BR "\-c, \-\-channel NAME"
Apply on channel name.
.TP
.BR "\-k, \-\-kernel"
Apply for the kernel tracer
.TP
.BR "\-u, \-\-userspace"
Apply for the user-space tracer
.TP
.BR "\-t, \-\-type TYPE"
Context type. You can repeat this option on the command line. Please
use "lttng add-context \-h" to list all available types.
.RE
.PP

.PP
\fBcalibrate\fP [OPTIONS]
.RS
Quantify LTTng overhead

The LTTng calibrate command can be used to find out the combined average
overhead of the LTTng tracer and the instrumentation mechanisms used. This
overhead can be calibrated in terms of time or using any of the PMU performance
counter available on the system.

For now, the only calibration implemented is that of the kernel function
instrumentation (kretprobes).

* Calibrate kernel function instrumentation

Let's use an example to show this calibration. We use an i7 processor with 4
general-purpose PMU registers. This information is available by issuing dmesg,
looking for "generic registers".

This sequence of commands will gather a trace executing a kretprobe hooked on
an empty function, gathering PMU counters LLC (Last Level Cache) misses
information (see lttng add-context \-\-help to see the list of available PMU
counters).

.nf
# lttng create calibrate-function
# lttng enable-event calibrate \-\-kernel \\
	\-\-function lttng_calibrate_kretprobe
# lttng add-context \-\-kernel \-t perf:cpu:LLC-load-misses \\
	\-t perf:cpu:LLC-store-misses \\
	\-t perf:cpu:LLC-prefetch-misses
# lttng start
# for a in $(seq 1 10); do \\
        lttng calibrate \-\-kernel \-\-function;
  done
# lttng destroy
# babeltrace $(ls \-1drt ~/lttng-traces/calibrate-function-* \\
	| tail \-n 1)
.fi

The output from babeltrace can be saved to a text file and opened in a
spreadsheet (e.g. oocalc) to focus on the per-PMU counter delta between
consecutive "calibrate_entry" and "calibrate_return" events. Note that these
counters are per-CPU, so scheduling events would need to be present to account
for migration between CPU. Therefore, for calibration purposes, only events
staying on the same CPU must be considered.

The average result, for the i7, on 10 samples:

.nf
                          Average     Std.Dev.
perf_LLC_load_misses:       5.0       0.577
perf_LLC_store_misses:      1.6       0.516
perf_LLC_prefetch_misses:   9.0      14.742
.fi

As we can notice, the load and store misses are relatively stable across runs
(their standard deviation is relatively low) compared to the prefetch misses.
We can conclude from this information that LLC load and store misses can be
accounted for quite precisely, but prefetches within a function seems to behave
too erratically (not much causality link between the code executed and the CPU
prefetch activity) to be accounted for.

.B OPTIONS:

.TP
.BR "\-h, \-\-help"
Show summary of possible options and commands.
.TP
.BR "\-k, \-\-kernel"
Apply for the kernel tracer
.TP
.BR "\-u, \-\-userspace"
Apply for the user-space tracer
.TP
.BR "\-\-function"
Dynamic function entry/return probe (default)
.RE
.PP

.PP
\fBcreate\fP [NAME] [OPTIONS]
.RS
Create tracing session.

A tracing session contains channel(s) which contains event(s). It is domain
agnostic, meaning that channels and events can be enabled for the
user-space tracer and/or the kernel tracer. It acts like a container
aggregating multiple tracing sources.

On creation, a \fB.lttngrc\fP file is created in your $HOME directory
containing the current session name. If NAME is omitted, a session name is
automatically created having this form: 'auto-yyyymmdd-hhmmss'.

If no \fB\-o, \-\-output\fP is specified, the traces will be written in
$HOME/lttng-traces.

The $HOME environment variable can be overridden by defining the environment
variable LTTNG_HOME. This is useful when the user running the commands has
a non-writeable home directory.

The session name MUST NOT contain the character '/'.

.B OPTIONS:

.TP
.BR "\-h, \-\-help"
Show summary of possible options and commands.
.TP
.BR "\-\-list-options"
Simple listing of options
.TP
.BR "\-o, \-\-output PATH"
Specify output path for traces
.TP
.BR "\-\-no-output"
Traces will not be output
.TP
.BR "\-\-snapshot"
Set the session in snapshot mode. Created in no-output mode and uses the
URL, if one is specified, as the default snapshot output.  Every channel will be set
in overwrite mode and with mmap output (splice not supported).
.TP
.BR "\-\-live [delay]"
Set the session exclusively in live mode. The parameter is the delay before
the data is flushed and streamed. The delay is expressed in microseconds by
default, but can also be suffixed with m (milliseconds) or s (seconds). The
live mode allows you to stream the trace and view it while it's being recorded
by any tracer. For that, you need a lttng-relayd and this session requires a
network URL (\-U or \-C/\-D). If no delay nor URL is provided, the default is
to use a timer value set to 1000000 and the network URL set to net://127.0.0.1.

To read a live session, you can use babeltrace(1) or the live streaming
protocol in doc/live-reading-protocol.txt. Here is an example:

.nf
$ lttng-relayd -o /tmp/lttng
$ lttng create --live 200m -U net://localhost
$ lttng enable-event -a --userspace
$ lttng start
.fi

After the start, you'll be able to read the events while they are being
recorded in /tmp/lttng.

.TP
.BR "\-\-shm-path PATH"

Path where shared memory holding buffers should be created. Useful
when used with PRAMFS or other persistent memory filesystems to extract
trace data in the event of a crash requiring a reboot.

See the \fBlttng-crash(1)\fP utility for more information on crash recovery.

.TP
.BR "\-U, \-\-set-url=URL"
Set URL for the consumer output destination. It is persistent for the
session lifetime. Redo the command to change it. This will set both data
and control URL for network.
.TP
.BR "\-C, \-\-ctrl-url=URL"
Set control path URL. (Must use -D also)
.TP
.BR "\-D, \-\-data-url=URL"
Set data path URL. (Must use -C also)
.PP
Using these options, each API call can be controlled individually. For
instance, \-C does not enable the consumer automatically. You'll need the \-e
option for that.

.B URL FORMAT:

proto://[HOST|IP][:PORT1[:PORT2]][/TRACE_PATH]

Supported protocols are (proto):
.TP
.BR "file://..."
Local filesystem full path.

.TP
.BR "net://..."
This will use the default network transport layer which is TCP for both
control (PORT1) and data port (PORT2). The default ports are
respectively 5342 and 5343. Note that net[6]:// is not yet supported.

.TP
.BR "tcp[6]://..."
Can only be used with -C and -D together

NOTE: IPv6 address MUST be enclosed in brackets '[]' (rfc2732)

.B EXAMPLES:

.nf
# lttng create -U net://192.168.1.42
.fi
Uses TCP and default ports for the given destination.

.nf
# lttng create -U net6://[fe80::f66d:4ff:fe53:d220]
.fi
Uses TCP, default ports and IPv6.

.nf
# lttng create s1 -U net://myhost.com:3229
.fi
Create session s1 and set its consumer to myhost.com on port 3229 for control.
.RE
.PP

.PP
\fBdestroy\fP [NAME] [OPTIONS]
.RS
Teardown tracing session

Free memory on the session daemon and tracer side. It's gone!

If NAME is omitted, the session name is taken from the .lttngrc file.

.B OPTIONS:

.TP
.BR "\-h, \-\-help"
Show summary of possible options and commands.
.TP
.BR "\-a, \-\-all"
Destroy all sessions
.TP
.BR "\-\-list-options"
Simple listing of options
.RE
.PP

.PP
\fBenable-channel\fP NAME[,NAME2,...] (\-k | \-u) [OPTIONS]
.RS
Enable tracing channel

To enable an event, you must enable both the event and the channel that
contains it.

If \fB\-s, \-\-session\fP is omitted, the session name is taken from the .lttngrc
file.

Exactly one of \-k or -u must be specified.

It is important to note that if a certain type of buffers is used, the session
will be set with that type and all other subsequent channel needs to have the
same type.

Note that once the session has been started and enabled on the tracer side,
it's not possible anymore to enable a new channel for that session.

.B OPTIONS:

.TP
.BR "\-h, \-\-help"
Show this help
.TP
.BR "\-\-list-options"
Simple listing of options
.TP
.BR "\-s, \-\-session NAME"
Apply on session name
.TP
.BR "\-k, \-\-kernel"
Apply to the kernel tracer
.TP
.BR "\-u, \-\-userspace"
Apply to the user-space tracer
.TP
.BR "\-\-discard"
Discard event when subbuffers are full (default)
.TP
.BR "\-\-overwrite"
Flight recorder mode: overwrites events when subbuffers are full. The
number of subbuffer must be 2 or more.
.TP
.BR "\-\-subbuf-size SIZE"
Subbuffer size in bytes {+k,+M,+G}.
(default UST uid: 131072, UST pid: 4096, kernel: 262144, metadata: 4096)
Rounded up to the next power of 2.

The minimum subbuffer size, for each tracer, is the max value between
the default above and the system page size. You can issue this command
to get the current page size on your system: \fB$ getconf PAGE_SIZE\fP
.TP
.BR "\-\-num-subbuf NUM"
Number of subbuffers. (default UST uid: 4, UST pid: 4, kernel: 4,
metadata: 2) Rounded up to the next power of 2.
.TP
.BR "\-\-switch-timer TIME"
Switch subbuffer timer interval in microseconds. Can also be suffixed with
m (milliseconds) or s (seconds).
(default UST uid: 0, UST pid: 0, kernel: 0, metadata: 0)
.TP
.BR "\-\-read-timer TIME"
Read timer interval in microseconds. Can also be suffixed with m (milliseconds)
or s (seconds).
(default UST uid: 0, UST pid: 0, kernel: 200000, metadata: 0)
.TP
.BR "\-\-output TYPE"
Channel output type. Possible values: mmap, splice
(default UST uid: mmap, UST pid: mmap, kernel: splice, metadata: mmap)
.TP
.BR "\-\-buffers-uid"
Use per UID buffer (\-u only). Buffers are shared between applications
that have the same UID.
.TP
.BR "\-\-buffers-pid"
Use per PID buffer (\-u only). Each application has its own buffers.
.TP
.BR "\-\-buffers-global"
Use shared buffer for the whole system (\-k only)
.TP
.BR "\-C, \-\-tracefile-size SIZE"
Maximum size of each tracefile within a stream (in bytes).
0 means unlimited. (default: 0)
Note: traces generated with this option may inaccurately report
discarded events as of CTF 1.8.
.TP
.BR "\-W, \-\-tracefile-count COUNT"
Used in conjunction with \-C option, this will limit the number of files
created to the specified count. 0 means unlimited. (default: 0)

.B EXAMPLES:

.nf
$ lttng enable-channel -k -C 4096 -W 32 chan1
.fi
For each stream, the maximum size of each trace file will be 4096 bytes and
there will be a maximum of 32 different files. The file count is appended after
the stream number as seen in the following example. The last trace file is
smaller than 4096 since it was not completely filled.

.nf
        ~/lttng-traces/[...]/chan1_0_0 (4096)
        ~/lttng-traces/[...]/chan1_0_1 (4096)
        ~/lttng-traces/[...]/chan1_0_2 (3245)
        ~/lttng-traces/[...]/chan1_1_0 (4096)
        ...
.fi

.nf
$ lttng enable-channel -k -C 4096
.fi
This will create trace files of 4096 bytes and will create new ones as long as
there is data available.
.RE
.PP

.PP
\fBenable-event\fP NAME[,NAME2,...] (\-k | \-u | \-j | \-l | \-p) [OPTIONS]
.RS
Enable tracing event

A tracing event is always assigned to a channel. If \fB\-c, \-\-channel\fP is
omitted, a default channel named '\fBchannel0\fP' is created and the event is

ChangeLog
2016-03-18 lttng-tools 2.8.0-rc1 (Awkward Moments Day)
	* Bump minor kernel tracer ABI version
	* Quote brewery names in release codename description
	* Fix: metadata regenerate, error if too many args
	* Fix: add missing sys/types.h header
	* Fix: build failure when __GLIBC_PREREQ is missing
	* configure.ac: beautify report
	* Update version codename and description to 2.8.0 - Isseki Nicho
	* doc/man: lttng.1.txt: sort commands in categories
	* Data pending message is printed when unnecessary
	* README.md: add CI/Coverity status badges
	* README.md: add lttng-crash in package contents
	* Show lttng-metadata man page in command's --help
	* Add an lttng-metadata man page
	* Remove untrack reference from snapshot man page
	* README.md: add optional dependency on man
	* doc/man: install lttng-health-check.3
	* lttng-relayd: show man page with --help and remove usage
	* doc/man: convert lttng-relayd(8) to AsciiDoc
	* lttng-crash.c: show man page with --help and remove usage
	* doc/man: convert lttng-crash(1) to AsciiDoc
	* lttng-sessiond: show man page with --help and remove usage
	* doc/man: convert lttng-sessiond(8) to AsciiDoc
	* README.md: bolden all dependencies and versions
	* README.md: add asciidoc and xmlto build dependencies
	* lttng.c: sort commands to get a nice list output
	* doc/man: add README.md
	* lttng: add `help` command
	* lttng.c: show man page with --help and remove usage
	* lttng.c: CMD_UNDEFINED is used for invalid args too
	* lttng: remove usage strings from commands
	* lttng: show man page when using command's --help
	* lttng add-context: add --list option
	* doc/man: remove lttng.1 (built from lttng.1.txt)
	* doc/man: create lttng-disable-event(1) and update/fix content
	* doc/man: create lttng-enable-event(1) and update/fix content
	* doc/man: create lttng-snapshot(1) and update/fix content
	* doc/man: create lttng-help(1) and update/fix content
	* doc/man: create lttng-status(1) and update/fix content
	* doc/man: create lttng-untrack(1) and update/fix content
	* doc/man: create lttng-track(1) and update/fix content
	* doc/man: create lttng-calibrate(1) and update/fix content
	* doc/man: create lttng-list(1) and update/fix content
	* doc/man: create lttng-add-context(1) and update/fix content
	* doc/man: create lttng-disable-channel(1) and update/fix content
	* doc/man: create lttng-enable-channel(1) and update/fix content
	* doc/man: create lttng-view(1) and update/fix content
	* doc/man: create lttng-version(1) and update/fix content
	* doc/man: create lttng-stop(1) and update/fix content
	* doc/man: create lttng-start(1) and update/fix content
	* doc/man: create lttng-load(1) and update/fix content
	* doc/man: create lttng-save(1) and update/fix content
	* doc/man: create lttng-set-session(1) and update/fix content
	* doc/man: create lttng-destroy(1) and update/fix content
	* doc/man: create lttng-create(1) and update/fix content
	* doc/man: add common OPTIONS head for command man pages
	* doc/man: add Makefile.am for generating man pages
	* .gitignore: ignore generated man page artifacts
	* doc/man: add XSL files for man pages
	* doc/man: add asciidoc.conf
	* configure.ac: check for asciidoc and xmlto
	* doc/man: lttng(1): update and standardize content
	* doc/man: convert lttng(1) to AsciiDoc
	* Add builtin modules support to kmod modprobe
	* Add comments to kmod functions
	* Add libkmod rmmod support
	* Move libkmod ifdef to beginning of file
	* Drop optional control modules load and unload
	* Fix: live test header endianness
	* Document the live detach command
	* Test the viewer detach command
	* Use of bare attribute directives instead of LTTNG_PACKED
	* Live: command to detach a viewer session
	* Contexts for RT debugging
	* Optional test for the metadata regeneration
	* Add MI output to the metadata regenerate command
	* Tests for metadata regenerate
	* Add TAGS files to .gitignore
	* Fix: systematic overwrite of union values on add context
	* Mismatching signature of ust_metadata_session_statedump() stub
	* Fix: Initialize events discarded and packet lost counters to zero
	* Assert on unknown UST buffer type
	* Command metadata regenerate
	* Store the relay protocol version in the consumer_output
	* Store the version of the tracer in the UID registry
	* Store the instance id and packet_seq_num in indexes
	* kernctl commands to extract the stream instance id
	* Account the lost packets in snapshot mode
	* Test: Add the lttng-runas worker process to the sessiond pids
	* Limit the scope of IFS overwriting
	* Do not overwrite IFS globally but only locally
	* Display discarded and lost events at destroy and stop
	* Fix: tests: use SIGSTOP for crash extraction test
	* Test fix: ignore stderr at every step of randstring
	* Fix: session_find_by_id can return NULL legitimately
	* Fix: mark ltt_sessions_ht_destroy as static
	* Docs: Missing locking assumptions in function headers
	* Test fix: ignore stderr when generating random string
	* Fix: hold session list lock during delete_ust_app
	* Test: prevent the spawning of a daemonized sessiond
	* Missing void in function signature
	* CLI: Show filter expression associated to a syscall
	* Fix: missing dereference when computing extended info position
	* Fix: Only save kernel enablers in session configuration
	* Fix: Only list kernel enablers when listing events
	* Fix: syscalls hash table leaks when listing kernel events
	* Initialize command header to zero
	* Rename filter string to filter expression in liblttng-ctl
	* Fix: use of unsigned variable to check for negative return
	* Clean-up: remove unreachable goto
	* Fix: missing static inline breaks --without-lttng-ust build
	* lttng enable-channel memsets lttng_channel to -1
	* Change padding type from array to fixed-width integer
	* Clean-up: missing space between cast operator and operand
	* Ensure UST channel output mode is LTTNG_UST_MMAP
	* Extract the lost packets and discarded events counters
	* Send extended channel payload to client

lttng/lttng.c
/*
 * Copyright (c)  2011 David Goulet <david.goulet@polymtl.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * as published by the Free Software Foundation; only version 2
 * of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#define _GNU_SOURCE
#include <getopt.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <config.h>

#include <lttng/lttng.h>

#include "cmd.h"
#include "conf.h"
#include "lttngerr.h"

/* Variables */
static char *progname;

int opt_quiet;
int opt_verbose;
static int opt_no_sessiond;
static char *opt_sessiond_path;

enum {
	OPT_SESSION_PATH,
	OPT_DUMP_OPTIONS,
	OPT_DUMP_COMMANDS,
};

/* Getopt options. No first level command. */
static struct option long_options[] = {
	{"help",             0, NULL, 'h'},
	{"group",            1, NULL, 'g'},
	{"verbose",          0, NULL, 'v'},
	{"quiet",            0, NULL, 'q'},
	{"no-sessiond",      0, NULL, 'n'},
	{"sessiond-path",    1, NULL, OPT_SESSION_PATH},
	{"list-options",     0, NULL, OPT_DUMP_OPTIONS},
	{"list-commands",    0, NULL, OPT_DUMP_COMMANDS},
	{NULL, 0, NULL, 0}
};

/* First level command */
static struct cmd_struct commands[] =  {
	{ "list", cmd_list},
	{ "create", cmd_create},
	{ "destroy", cmd_destroy},
	{ "start", cmd_start},
	{ "stop", cmd_stop},
	{ "enable-event", cmd_enable_events},
	{ "disable-event", cmd_disable_events},
	{ "enable-channel", cmd_enable_channels},
	{ "disable-channel", cmd_disable_channels},
	{ "add-context", cmd_add_context},
	{ "set-session", cmd_set_session},
	{ "version", cmd_version},
	{ "calibrate", cmd_calibrate},
	{ NULL, NULL}	/* Array closure */
};

static void usage(FILE *ofp)
{
	fprintf(ofp, "LTTng Trace Control " VERSION"\n\n");
	fprintf(ofp, "usage: lttng [options] <command>\n");
	fprintf(ofp, "\n");
	fprintf(ofp, "Options:\n");
	fprintf(ofp, "  -h, --help             Show this help\n");
	fprintf(ofp, "  -g, --group NAME       Unix tracing group name. (default: tracing)\n");
	fprintf(ofp, "  -v, --verbose          Verbose mode\n");
	fprintf(ofp, "  -q, --quiet            Quiet mode\n");
	fprintf(ofp, "  -n, --no-sessiond      Don't spawn a session daemon\n");
	fprintf(ofp, "      --sessiond-path    Session daemon full path\n");
	fprintf(ofp, "      --list-options     Simple listing of lttng options\n");
	fprintf(ofp, "      --list-commands    Simple listing of lttng commands\n");
	fprintf(ofp, "\n");
	fprintf(ofp, "Commands:\n");
	fprintf(ofp, "    add-context     Add context to event or/and channel\n");
	fprintf(ofp, "    calibrate       Quantify LTTng overhead\n");
	fprintf(ofp, "    create          Create tracing session\n");
	fprintf(ofp, "    destroy         Teardown tracing session\n");
	fprintf(ofp, "    enable-channel  Enable tracing channel\n");
	fprintf(ofp, "    enable-event    Enable tracing event\n");
	fprintf(ofp, "    disable-channel Disable tracing channel\n");
	fprintf(ofp, "    disable-event   Disable tracing event\n");
	fprintf(ofp, "    list            List possible tracing options\n");
	fprintf(ofp, "    set-session     Set current session name\n");
	fprintf(ofp, "    start           Start tracing\n");
	fprintf(ofp, "    stop            Stop tracing\n");
	fprintf(ofp, "    version         Show version information\n");
	fprintf(ofp, "\n");
	fprintf(ofp, "Please see the lttng(1) man page for full documentation.\n");
	fprintf(ofp, "See http://lttng.org for updates, bug reports and news.\n");
}

/*
 *  list_options
 *
 *  List options line by line. This is mostly for bash auto completion and to
 *  avoid difficult parsing.
 */
static void list_options(FILE *ofp)
{
	int i = 0;
	struct option *option = NULL;

	option = &long_options[i];
	while (option->name != NULL) {
		fprintf(ofp, "--%s\n", option->name);

		if (isprint(option->val)) {
			fprintf(ofp, "-%c\n", option->val);
		}

		i++;
		option = &long_options[i];
	}
}

/*
 *  list_commands
 *
 *  List commands line by line. This is mostly for bash auto completion and to
 *  avoid difficult parsing.
 */
static void list_commands(FILE *ofp)
{
	int i = 0;
	struct cmd_struct *cmd = NULL;

	cmd = &commands[i];
	while (cmd->name != NULL) {
		fprintf(ofp, "%s\n", cmd->name);
		i++;
		cmd = &commands[i];
	}
}

/*
 * clean_exit
 */
static void clean_exit(int code)
{
	DBG("Clean exit");
	exit(code);
}

/*
 *  sighandler
 *
 *  Signal handler for the daemon
 */
static void sighandler(int sig)
{
	switch (sig) {
		case SIGTERM:
			DBG("SIGTERM catched");
			clean_exit(EXIT_FAILURE);
			break;
		case SIGCHLD:
			/* Notify is done */
			DBG("SIGCHLD catched");
			break;
		default:
			DBG("Unknown signal %d catched", sig);
			break;
	}

	return;
}

/*
 *  set_signal_handler
 *
 *  Setup signal handler for SIGCHLD and SIGTERM.
 */
static int set_signal_handler(void)
{
	int ret = 0;
	struct sigaction sa;
	sigset_t sigset;

	if ((ret = sigemptyset(&sigset)) < 0) {
		perror("sigemptyset");
		goto end;
	}

	sa.sa_handler = sighandler;
	sa.sa_mask = sigset;
	sa.sa_flags = 0;
	if ((ret = sigaction(SIGCHLD, &sa, NULL)) < 0) {
		perror("sigaction");
		goto end;
	}

	if ((ret = sigaction(SIGTERM, &sa, NULL)) < 0) {
		perror("sigaction");
		goto end;
	}

end:
	return ret;
}

/*
 *  handle_command
 *
 *  Handle the full argv list of a first level command. Will find the command
 *  in the global commands array and call the function callback associated.
 *
 *  If command not found, return -1
 *  else, return function command error code.
 */
static int handle_command(int argc, char **argv)
{
	int i = 0, ret;
	struct cmd_struct *cmd;

	if (*argv == NULL) {
		ret = CMD_SUCCESS;
		goto end;
	}

	cmd = &commands[i];
	while (cmd->func != NULL) {
		/* Find command */
		if (strcmp(argv[0], cmd->name) == 0) {
			ret = cmd->func(argc, (const char**) argv);
			switch (ret) {
			case CMD_ERROR:
				ERR("Command error");
				break;
			case CMD_NOT_IMPLEMENTED:
				ERR("Options not implemented");
				break;
			case CMD_UNDEFINED:
				ERR("Undefined command");
				break;
			case CMD_FATAL:
				ERR("Fatal error");
				break;
			}
			goto end;
		}
		i++;
		cmd = &commands[i];
	}

	/* Command not found */
	ret = -1;

end:
	return ret;
}

/*
 *  spawn_sessiond
 *
 *  Spawn a session daemon by forking and execv.
 */
static int spawn_sessiond(char *pathname)
{
	int ret = 0;
	pid_t pid;

	MSG("Spawning a session daemon");
	pid = fork();
	if (pid == 0) {
		/*
		 * Spawn session daemon and tell
		 * it to signal us when ready.
		 */
		execlp(pathname, "lttng-sessiond", "--sig-parent", "--quiet", NULL);
		/* execlp only returns if error happened */
		if (errno == ENOENT) {
			ERR("No session daemon found. Use --sessiond-path.");
		} else {
			perror("execlp");
		}
		kill(getppid(), SIGTERM);	/* unpause parent */
		exit(EXIT_FAILURE);
	} else if (pid > 0) {
		/* Wait for lttng-sessiond to start */
		pause();
		goto end;
	} else {
		perror("fork");
		ret = -1;
		goto end;
	}

end:
	return ret;
}

/*
 *  check_sessiond
 *
 *  Check if the session daemon is available using
 *  the liblttngctl API for the check. If not, try to
 *  spawn a daemon.
 */
static int check_sessiond(void)
{
	int ret;
	char *pathname = NULL, *alloc_pathname = NULL;

	ret = lttng_session_daemon_alive();
	if (ret == 0) {	/* not alive */
		/* Try command line option path */
		if (opt_sessiond_path != NULL) {
			ret = access(opt_sessiond_path, F_OK | X_OK);
			if (ret < 0) {
				ERR("No such file: %s", opt_sessiond_path);
				goto end;
			}
			pathname = opt_sessiond_path;
		} else {
			/* Try LTTNG_SESSIOND_PATH env variable */
			pathname = getenv(LTTNG_SESSIOND_PATH_ENV);
		}

		/* Let's rock and roll */
		if (pathname == NULL) {
			ret = asprintf(&alloc_pathname, INSTALL_BIN_PATH "/lttng-sessiond");
			if (ret < 0) {
				perror("asprintf spawn sessiond");
				goto end;
			}
			pathname = alloc_pathname;
		}

		ret = spawn_sessiond(pathname);
		free(alloc_pathname);
		if (ret < 0) {
			ERR("Problem occurs when starting %s", pathname);
			goto end;
		}
	}

end:
	return ret;
}

/*
 * Check for the "help" option in the argv. If found, return 1 else return 0.
 */
static int check_help_command(int argc, char **argv)
{
	int i;

	for (i = 0; i < argc; i++) {
		if ((strncmp(argv[i], "-h", 2) == 0) ||
				strncmp(argv[i], "--h", 3) == 0) {
			return 1;
		}
	}

	return 0;
}

/*
 *  parse_args
 *
 *  Parse command line arguments.
 *  Return 0 if OK, else -1
 */
static int parse_args(int argc, char **argv)
{
	int opt, ret;

	if (argc < 2) {
		usage(stderr);
		clean_exit(EXIT_FAILURE);
	}

	while ((opt = getopt_long(argc, argv, "+hnvqg:", long_options, NULL)) != -1) {
		switch (opt) {
		case 'h':
			usage(stderr);
			goto error;
		case 'v':
			opt_verbose += 1;
			break;
		case 'q':
			opt_quiet = 1;
			break;
		case 'g':
			lttng_set_tracing_group(optarg);
			break;
		case 'n':
			opt_no_sessiond = 1;
			break;
		case OPT_SESSION_PATH:
			opt_sessiond_path = strdup(optarg);
			break;
		case OPT_DUMP_OPTIONS:
			list_options(stdout);
			ret = 0;
			goto error;
		case OPT_DUMP_COMMANDS:
			list_commands(stdout);
			ret = 0;
			goto error;
		default:
			usage(stderr);
			goto error;
		}
	}

	/* If both options are specified, quiet wins */
	if (opt_verbose && opt_quiet) {
		opt_verbose = 0;
	}

	/* Spawn session daemon if needed */
	if (opt_no_sessiond == 0 && check_help_command(argc, argv) == 0 &&
			(check_sessiond() < 0)) {
		goto error;
	}

	/* No leftovers, print usage and quit */
	if ((argc - optind) == 0) {
		usage(stderr);
		goto error;
	}

	/* 
	 * Handle leftovers which is a first level command with the trailing
	 * options.
	 */
	ret = handle_command(argc - optind, argv + optind);
	if (ret < 0) {
		if (ret == -1) {
			usage(stderr);
		} else {
			ERR("%s", lttng_strerror(ret));
		}
		goto error;
	}

	return 0;

error:
	return -1;
}


/*
 *  main
 */
int main(int argc, char *argv[])
{
	int ret;

	progname = argv[0] ? argv[0] : "lttng";

	/* For Mathieu Desnoyers aka Dr Tracing */
	if (strncmp(progname, "drtrace", 7) == 0) {
		MSG("%c[%d;%dmWelcome back Dr Tracing!%c[%dm\n\n", 27,1,33,27,0);
	}

	ret = set_signal_handler();
	if (ret < 0) {
		clean_exit(ret);
	}

	ret = parse_args(argc, argv);
	if (ret < 0) {
		clean_exit(EXIT_FAILURE);
	}

	return 0;
}
